game-core/src/main/java/games/strategy/triplea/util/MovableUnitsFilter.java

```diff
@@ -0,0 +1,202 @@
+        best = best.subList(1, best.size());
+        lastResult = validateMoveWithDependents(best, dependentUnits, transportsToLoad);
+      }
+    }
+
+    return makeResult(allUnitsResult, lastResult);
+  }
+
+  private Result makeResult(
+      final MoveValidationResult allUnitsResult,
+      final MoveValidationResultWithDependents lastResult) {
+    @Nullable String errorMessage = null;
+    @Nullable String warningMessage = null;
+    if (!allUnitsResult.isMoveValid()) {
+      @Nullable String message = allUnitsResult.getError();
+      if (message == null) {
+        message = allUnitsResult.getDisallowedUnitWarning(0);
+      }
+      if (message == null) {
+        message = allUnitsResult.getUnresolvedUnitWarning(0);
+      }
+      if (!lastResult.getResult().isMoveValid()) {
+        errorMessage = message;
+      } else {
+        warningMessage = message;
+      }
+    }
+    return new Result(lastResult.getUnitsWithDependents(), errorMessage, warningMessage);
+  }
+
+  private Collection<Unit> getPossibleTransportsToLoad(
+      final Collection<Unit> units,
+      final Map<Unit, Collection<Unit>> dependentUnits,
+      final Route route) {
+    // TODO kev check for already loaded airTransports
+    if (MoveValidator.isLoad(units, dependentUnits, route, data, player)) {
+      final UnitCollection unitsAtEnd = route.getEnd().getUnitCollection();
+      return unitsAtEnd.getMatches(Matches.unitIsTransport().and(Matches.alliedUnit(player, data)));
+    }
+    return List.of();
+  }
+
+  private Comparator<Unit> getUnitComparator(final List<Unit> units) {
+    // sort units based on which transports are allowed to unload
+    if (route.isUnload() && units.stream().anyMatch(Matches.unitIsLand())) {
+      return UnitComparator.getUnloadableUnitsComparator(units, route, player);
+    } else {
+      return UnitComparator.getMovableUnitsComparator(units, route);
+    }
+  }
+
+  private MoveValidationResultWithDependents validateMoveWithDependents(
+      final List<Unit> units,
+      final Map<Unit, Collection<Unit>> dependentUnits,
+      final Collection<Unit> transportsToLoad) {
+    final List<Unit> unitsWithDependents = addMustMoveWith(units, dependentUnits);
+    final MoveValidationResult result;
+    data.acquireReadLock();
+    try {
+      final Map<Unit, Unit> unitsToTransports =
+          transportsToLoad.isEmpty()
+              ? Map.of()
+              : TransportUtils.mapTransports(route, units, transportsToLoad);
+      final MoveDescription move =
+          new MoveDescription(unitsWithDependents, route, unitsToTransports, dependentUnits);
+      result =
+          AbstractMoveDelegate.validateMove(moveType, move, player, nonCombat, undoableMoves, data);
+    } finally {
+      data.releaseReadLock();
+    }
+
+    return new MoveValidationResultWithDependents(
+        result, result.isMoveValid() ? unitsWithDependents : List.of());
+  }
+
+  private List<Unit> addMustMoveWith(
```

Comment:
@DanVanAtta I'm moving this method. I actually simplified it slightly by removing an extra condition - i.e. can just check whether .get() returned null instead of checking membership, but still codeclimate complains it's too complex.

Given the type of code in TripleA, this seems an unreasonable complaint. I think this method is of reasonable complexity for TripleA and I'd be happy if most of the code of TripleA is of this complexity. (Most is much more complex.)

Perhaps we should change the codeclimate threshold? The alternative is to make a ton of small helper methods which introduces its own complexity (i.e. makes classes longer and hard to understand how things fit together).

@DanVanAtta Do you have a better suggestion on what to do here? I mean, certainly we could change helpers...

For example, some ideas I see:
  - Make bestWithDependents a Set, which can avoid the contains() call, but then we have to convert to ArrayList on return.
  - We could make MustMoveWithDetails have some method to do some of this functionality. For example, instead of calling of getMustMoveWith().get(u), we could call a method that returns the units it must move with - never null - which can avoid the null check.

But even with all that, it's not obvious this will reduce the complexity enough. Is there a way to test codeclimate locally - or do I need to upload new patch sets and wait for codeclimate to comment on the PR?

Reply:
Also, the logic in this function doesn't make sense to me, particularly this part:

```
      final Collection<Unit> mustMoveWith = mustMoveWithDetails.getMustMoveWith().get(u);
      if (mustMoveWith != null) {
        for (final Unit m : mustMoveWith) {
          if (!bestWithDependents.contains(m)) {
            bestWithDependents.addAll(mustMoveWith);
          }
        }
      }
```

It seems ```mustMovewith``` gets added multiple types for each unit in ```mustMoveWith``` - which seems incorrect. It seems the correct thing is to have the inner statement be ```bestWithDependents.add(m);``` - but this PR is not meant to change behavior here... so....

Reply:
For now, I've simplified this function so that the null check is avoided and also reduced the number of params needed for MoveValidator.getMustMoveWith(), which should hopefully satisfy codeclimate...